### PR TITLE
BrowserStack Integration

### PR DESCRIPTION
**Pull request by jackdwalker**
Thanks for submitting a PR! Please fill in this template where appropriate:

## Category

Tests

## Feature/Issue Description

This integration will allow BeEF to run any components of the test suite that require hooking/victim browser interaction against a matrix of Browser/Operating Systems and their different versions. 

This will help ensure that future code added to BeEF work on both new and old systems, which is vital considering the nature of BeEF's functionality. Many of the older systems are those are would be most susceptible and we want to ensure that we have BeEF working against such environments.

## Describe Changes 

Specific tests now send encrypted traffic through to BrowserStack and the Selenium web-driver will run within their environment rather than on the host system of the Travis box running the test suite.

These tests are DISABLED by default when running the standard test command:

```bash
bundle exec rake
```

The tests will only run when Travis-CI kicks off the test suite so functionally for users running testing locally there should be no change.

## Test Cases

No new tests, although there were many modifications to the tests utilising a dummy victim to allow for this integration. Some cleanup was done on the code of those tests also.

## Relevant Wiki Page

I have a Wiki page drafted up in Markdown, ready for the merge of this PR that details the baseline level of compatibility tested by this integration.

### CODE REVIEW

**Comment by jackdwalker** [comment]
@jcrew99 @DeezyE - could you please both look over this before we merge tomorrow morning.

File: spec/beef/core/main/autorun_engine/autorun_engine_spec.rb [hunk 1]
```diff
@@ -6,103 +6,131 @@
...
+      sleep 1 until wait.until { @driver.execute_script('return window.beef.session.get_hook_session_id().length') > 0 }
+
+      @session = @driver.execute_script('return window.beef.session.get_hook_session_id()')
+    rescue StandardError => e
+      print_info "Exception: #{e}"
+      print_info "Exception Class: #{e.class}"
+      print_info "Exception Message: #{e.message}"
+      print_info "Exception Stack Trace: #{e.backtrace}"
+      if @driver.execute_script('return window.beef.session.get_hook_session_id().length').nil? &&
+         e.class == NoMethodError
```
Comment (DeezyE) [inline spec/beef/core/main/autorun_engine/autorun_engine_spec.rb:110]: as per our discussion

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -22,6 +26,24 @@
 ENV['RACK_ENV'] ||= 'test'
 ARGV = []
 
+## BrowserStack config
+
```
Comment (jcrew99) [inline spec/spec_helper.rb:30]: @jackdwalker  This doesn't seem ideal but if it works it can be picked up at some point later

File: spec/support/constants.rb [hunk 3]
```diff
@@ -6,8 +6,8 @@
 BEEF_TEST_DIR = "/tmp/beef-test/"
 
 # General constants
-ATTACK_DOMAIN = "127.0.0.1"
```
Comment (jcrew99) [inline spec/support/constants.rb:9]: Why was this change necessary? Has everything that depends on it been changed?

File: spec/spec_helper.rb (hunk 2 above)
Comment (jcrew99) [reply spec/spec_helper.rb:30]: To be clear, i am referring to the monkey patching. i think its just worth noting and possibly making a plan to action it

File: local.log [hunk 4]
```diff
@@ -0,0 +1,2117 @@
+
```
[the remark this answers is not in the document]
Comment (jackdwalker) [reply local.log:1]: I agree this should be removed.

File: spec/support/constants.rb (hunk 3 above)
Comment (jackdwalker) [reply spec/support/constants.rb:9]: @jcrew99 - it helped prevent a BrowserStack exception, that occasionally occurred when using browserstack_local gem tunnelling + then setting your domain as 127.0.0.1 instead of localhost.

It's the same thing. There was no dependencies that I could identify that require ATTACK_DOMAIN/VICTIM_DOMAIN to be the exact way that they were. They just both need to refer to the local domain, and be different. Please see the documentation @ https://github.com/beefproject/beef/wiki/BeEF-Testing.

It can be changed back with the way we are now handling exceptions, although I do not anticipate it will cause any issues, so I don't think it's worth delaying the merge by another 6 hours.

File: spec/spec_helper.rb (hunk 2 above)
Comment (jackdwalker) [reply spec/spec_helper.rb:30]: @jcrew99 - that is not my code. That is in the existing BeEF master. Feel free to open an issue on it though.

File: local.log (hunk 4 above)
[the remark this answers is not in the document]
Comment (jackdwalker) [reply local.log:1]: @jcrew99 agreed. I'll add it to the gitignore.

File: core/main/client/browser.js [hunk 5]
```diff
@@ -109,7 +109,7 @@ beef.browser = {
      * @example: beef.browser.isIE10()
      */
     isIE10: function () {
-        return !!window.XMLHttpRequest && !window.chrome && !window.opera && !!document.documentMode && !window.XDomainRequest && !!window.performance && typeof navigator.msMaxTouchPoints !== "undefined";
+        return !!window.XMLHttpRequest && !window.chrome && !window.opera && !!document.documentMode && !!window.XDomainRequest && !!window.performance && typeof navigator.msMaxTouchPoints !== "undefined";
```
[the remark this answers is not in the document]
Comment (jackdwalker) [reply core/main/client/browser.js:112]: @jcrew99 - !!/double-bang operator is effectively not, not. If something evaluates to nil, in this case window.XMLHttpRequest, !window.XMLHttpRequest will evaluate to false (because nil is falsy), and so !!window.XMLHttpRequest will equal true, because !false is the same as true.

Practically this means that we can use !!functionHere return true when a function doesn't exist in a browser. In this case, part of the IE10 fingerprinting is to check whether or not the window.XMLHttpRequest function exists. We need this check to evaluate to true because we want the function isIE10() to return true when that function is missing among a bunch of other checks. I'd suggest reading:

https://medium.com/@edplatomail/js-double-bang-or-the-not-operator-part-40e55d089bf0

File: .travis.yml [hunk 6]
```diff
@@ -1,28 +1,93 @@
...
 rvm:
-  - 2.5.3
-  - 2.6.0
-  - 2.6.3
-  - 2.6.5
-  - 2.7.0
-notifications:
-  email:
-    recipients:
-      - wade@bindshell.net
```
[the remark this answers is not in the document]
Comment (jackdwalker) [reply .travis.yml:11]: @jcrew99 - he hasn't mentioned anything to me. We can remove it in a later PR if it's bothering him. I think he'd rather the branch be merged than wait 6 hours just to remove him from emails.

File: local.log (hunk 4 above)
[the remark this answers is not in the document]
Comment (jackdwalker) [reply local.log:1]: @jcrew99 I can't even find this on my local. It must be being generated. I might just merge and add it to the .gitignore in a separate PR.

File: core/main/client/browser.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (jcrew99) [reply core/main/client/browser.js:112]: Sounds good, i wasn't sure if it was wanted to be true instead  of false. Thanks for the reading suggestion, good to go through!

File: spec/spec_helper.rb (hunk 2 above)
Comment (jcrew99) [reply spec/spec_helper.rb:30]: ahaha cool

File: .travis.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (jcrew99) [reply .travis.yml:11]: Does this not remove him from the notifications?

File: local.log (hunk 4 above)
[the remark this answers is not in the document]
Comment (jcrew99) [reply local.log:1]: Sounds good

File: .travis.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (jcrew99) [reply .travis.yml:11]: Okay sure, just add it back in next one
